beacon-chain/blockchain/fork_choice.go

```diff
@@ -105,24 +115,50 @@ func (c *ChainService) updateFFGCheckPts(state *pb.BeaconState) error {
 
 // ApplyForkChoiceRule determines the current beacon chain head using LMD GHOST as a block-vote
 // weighted function to select a canonical head in Ethereum Serenity.
-func (c *ChainService) ApplyForkChoiceRule(ctx context.Context, block *pb.BeaconBlock, computedState *pb.BeaconState) error {
+func (c *ChainService) ApplyForkChoiceRule(ctx context.Context, block *pb.BeaconBlock, postState *pb.BeaconState) error {
 	ctx, span := trace.StartSpan(ctx, "beacon-chain.blockchain.ApplyForkChoiceRule")
 	defer span.End()
-	h, err := hashutil.HashBeaconBlock(block)
+	log.Info("Applying LMD-GHOST Fork Choice Rule")
+
+	attestationTargets, err := c.attestationTargets(postState)
 	if err != nil {
-		return fmt.Errorf("could not tree hash incoming block: %v", err)
+		return fmt.Errorf("could not retrieve attestation target: %v", err)
 	}
-	// TODO(#1307): Use LMD GHOST as the fork-choice rule for Ethereum Serenity.
-	// TODO(#674): Handle chain reorgs.
-	if err := c.beaconDB.UpdateChainHead(block, computedState); err != nil {
-		return fmt.Errorf("failed to update chain: %v", err)
+	justifiedHead, err := c.beaconDB.JustifiedBlock()
+	if err != nil {
+		return err
+	}
+	head, err := c.lmdGhost(justifiedHead, postState, attestationTargets)
+	if err != nil {
+		return fmt.Errorf("could not run fork choice: %v", err)
 	}
-	log.WithField("blockRoot", fmt.Sprintf("0x%x", h)).Info("Chain head block and state updated")
+	if head.Slot != block.Slot {
+		log.Warnf("Reorg happened, last processed block at slot %d, new head block at slot %d",
+			block.Slot-params.BeaconConfig().GenesisSlot, head.Slot-params.BeaconConfig().GenesisSlot)
 
-	if err := c.saveHistoricalState(computedState); err != nil {
-		log.Errorf("Could not save new historical state: %v", err)
+		// Only regenerate head state if there was a reorg.
+		postState, err = stategenerator.GenerateStateFromBlock(c.ctx, c.beaconDB, head.Slot)
+		if err != nil {
+			return fmt.Errorf("could not gen state: %v", err)
+		}
+
+		if postState.Slot != postState.Slot {
```

Comment:
this if condition looks weird wouldn't they always be the same?

Reply:
: looks like a search and replace causality where the parameter name was renamed to postState and we didnt have good enough test coverage for a test to fail